File: frontend/js/views/car_affordability.js
// Car affordability — the 20/4/10 rule worked back to a max car price, via
// /calc/car-affordability. Live as you type.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';
import * as enh from '../calc_enhance.js';

const money = (n) => '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 });
const VIEW = 'car-affordability';
let lastReport = null;
let lastBody = null;

export async function renderCarAffordability(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.caraff.h1.title">// CAR AFFORDABILITY</span></h1>
        <p class="muted small" data-i18n="view.caraff.hint.intro">
            The 20/4/10 rule of thumb — put 20% down, finance for no more than 4 years, and keep
            total vehicle spending under 10% of gross income. This works back from the income cap to
            the most car you can afford. Updates as you type.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.caraff.h2.inputs">Your budget</h2>
            <form id="caraff-form" class="inline-form">
                <label><span data-i18n="view.caraff.label.income">Annual income ($)</span>
                    <input type="number" step="0.01" min="0" name="annual_income_usd" value="60000" required></label>
                <label><span data-i18n="view.caraff.label.down">Down payment (%)</span>
                    <input type="number" step="1" min="0" max="99" name="down_payment_pct" value="20" required></label>
                <label><span data-i18n="view.caraff.label.term">Loan term (months)</span>
                    <input type="number" step="1" min="1" name="loan_term_months" value="48" required></label>
                <label><span data-i18n="view.caraff.label.apr">APR (%)</span>
                    <input type="number" step="0.01" min="0" name="apr_pct" value="6" required></label>
                <label><span data-i18n="view.caraff.label.pct">Max % of gross income</span>
                    <input type="number" step="0.1" min="0" name="max_payment_pct_of_income" value="10" required></label>
                <label><span data-i18n="view.caraff.label.insfuel">Insurance + fuel / mo ($)</span>
                    <input type="number" step="0.01" min="0" name="insurance_fuel_monthly_usd" value="0"></label>
            </form>
            <div id="caraff-tools" class="ce-toolbar"></div>
            <button type="button" id="caraff-sens-btn" class="ce-tool" data-i18n="calc.enh.sens.run">▦ Sensitivity</button>
            <div id="caraff-sens" class="ce-sens"></div>
        </div>
        <div id="caraff-result" class="lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#caraff-form');
    enh.prefillForm(form, enh.readHashInputs());
    const readBody = () => {
        const fd = new FormData(form);
        return {
            annual_income_usd: Number(fd.get('annual_income_usd')) || 0,
            down_payment_pct: Number(fd.get('down_payment_pct')) || 0,
            loan_term_months: Number(fd.get('loan_term_months')) || 0,
            apr_pct: Number(fd.get('apr_pct')) || 0,
            max_payment_pct_of_income: Number(fd.get('max_payment_pct_of_income')) || 0,
            insurance_fuel_monthly_usd: Number(fd.get('insurance_fuel_monthly_usd')) || 0,
        };
    };
    const generate = async () => {
        const body = readBody();
        try {
            const r = await api.calcCarAffordability(body);
            if (!viewIsCurrent(tok)) return;
            lastReport = r; lastBody = body;
            renderResult(mount, r, body, tok);
        } catch (err) {
            showToast(err.message || t('view.caraff.toast.error'), { level: 'error' });
        }
    };
    enh.mountToolbar(mount.querySelector('#caraff-tools'), { viewId: VIEW, getInputs: () => lastBody || readBody(), getRows: () => reportRows(lastReport), filename: 'car-affordability.csv' });
    mount.querySelector('#caraff-sens-btn').addEventListener('click', () => runSens(mount, readBody(), tok));
    form.addEventListener('input', debounce(generate, 250));
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function reportRows(r) {
    if (!r) return [];
    return [
        ['metric', 'value'],
        ['max_car_price_usd', r.max_car_price_usd],
        ['max_loan_usd', r.max_loan_usd],
        ['monthly_payment_budget_usd', r.monthly_payment_budget_usd],
        ['monthly_transport_budget_usd', r.monthly_transport_budget_usd],
        ['down_payment_needed_usd', r.down_payment_needed_usd],
    ];
}

async function renderResult(mount, r, body, tok) {
    const el = mount.querySelector('#caraff-result');
    // Line chart: max car price as APR sweeps 3% -> 12% (higher rate buys less car).
    const xs = enh.linspace(3, 12, 13);
    const pts = await Promise.all(xs.map(async (apr) => {
        const rr = await api.calcCarAffordability({ ...body, apr_pct: apr });
        return { x: apr, y: rr ? rr.max_car_price_usd / 1000 : NaN };
    }));
    if (!viewIsCurrent(tok)) return;
    const chart = enh.svgLineChart(pts, { xlabel: 'APR %', ylabel: 'max price $k' });
    el.innerHTML = `
        <div class="chart-panel">
            <h2 data-i18n="view.caraff.h2.result">What you can afford</h2>
            <div class="cards">
                <div class="card pos"><div class="label" data-i18n="view.caraff.card.price">Max car price</div>
                    <div class="value pos">${money(r.max_car_price_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.caraff.card.loan">Max loan</div>
                    <div class="value">${money(r.max_loan_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.caraff.card.payment">Payment budget</div>
                    <div class="value">${money(r.monthly_payment_budget_usd)}</div></div>
            </div>
            ${chart}
            <table class="data-table">
                <tbody>
                    <tr><td data-i18n="view.caraff.row.budget">Transport budget / mo</td><td>${money(r.monthly_transport_budget_usd)}</td></tr>
                    <tr><td data-i18n="view.caraff.row.payment">Payment budget / mo</td><td>${money(r.monthly_payment_budget_usd)}</td></tr>
                    <tr><td data-i18n="view.caraff.row.loan">Max loan</td><td>${money(r.max_loan_usd)}</td></tr>
                    <tr><td data-i18n="view.caraff.row.down">Down payment needed</td><td>${money(r.down_payment_needed_usd)}</td></tr>
                    <tr class="emph"><td data-i18n="view.caraff.row.price">Max car price</td><td>${money(r.max_car_price_usd)}</td></tr>
                </tbody>
            </table>
        </div>
    `;
    applyUiI18n(el);
}

async function runSens(mount, base, tok) {
    const panel = mount.querySelector('#caraff-sens');
    panel.innerHTML = `<p class="muted small" data-i18n="calc.enh.sens.running">Computing…</p>`; applyUiI18n(panel);
    // x: APR 3% -> 12%; y: annual income 0.6x -> 1.4x. Output: max car price.
    const inc = base.annual_income_usd || 60000;
    const xVals = enh.linspace(3, 12, 5);
    const yVals = enh.linspace(inc * 0.6, inc * 1.4, 5);
    const { cells } = await enh.runSensitivity({ base, xKey: 'apr_pct', yKey: 'annual_income_usd', xVals, yVals, compute: (b) => api.calcCarAffordability(b), pick: (r) => (r ? r.max_car_price_usd : null) });
    if (!viewIsCurrent(tok)) return;
    panel.innerHTML = enh.renderSensitivityTable({ xVals, yVals, cells, fmt: (v) => (v == null ? '—' : '$' + Math.round(v / 1000) + 'k'), xfmt: (v) => v.toFixed(1) + '%', yfmt: (v) => '$' + Math.round(v / 1000) + 'k', xName: t('view.caraff.label.apr') || 'APR', yName: t('view.caraff.label.income') || 'Income' });
}

File: crates/traderview-core/src/car_affordability.rs
//! Car affordability — the 20/4/10 rule of thumb: put 20% down, finance for no
//! more than 4 years, and keep total vehicle spending under 10% of gross income.
//!
//! Working backward from the income cap to a price:
//!
//! ```text
//! monthly budget = 10% × gross monthly income − insurance/fuel
//! max loan       = present value of that payment over the term
//! max price      = max loan / (1 − down payment %)
//! ```

use serde::{Deserialize, Serialize};

fn d_down() -> f64 {
    20.0
}
fn d_term() -> f64 {
    48.0
}
fn d_pct() -> f64 {
    10.0
}

#[derive(Debug, Clone, Deserialize)]
pub struct CarAffordInput {
    pub annual_income_usd: f64,
    #[serde(default = "d_down")]
    pub down_payment_pct: f64,
    #[serde(default = "d_term")]
    pub loan_term_months: f64,
    pub apr_pct: f64,
    /// Cap on monthly vehicle spending as a percent of gross income.
    #[serde(default = "d_pct")]
    pub max_payment_pct_of_income: f64,
    /// Monthly insurance + fuel, carved out of the budget before the payment.
    #[serde(default)]
    pub insurance_fuel_monthly_usd: f64,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct CarAffordResult {
    /// 10% of gross monthly income.
    pub monthly_transport_budget_usd: f64,
    /// Budget left for the loan payment after insurance/fuel.
    pub monthly_payment_budget_usd: f64,
    /// Largest loan that payment supports.
    pub max_loan_usd: f64,
    /// Largest car price (loan grossed up for the down payment).
    pub max_car_price_usd: f64,
    /// Down payment required on that price.
    pub down_payment_needed_usd: f64,
}

fn present_value(pmt: f64, i: f64, n: f64) -> f64 {
    if pmt <= 0.0 || n <= 0.0 {
        return 0.0;
    }
    if i.abs() < 1e-12 {
        pmt * n
    } else {
        pmt * (1.0 - (1.0 + i).powf(-n)) / i
    }
}

pub fn analyze(input: &CarAffordInput) -> CarAffordResult {
    let monthly_gross = input.annual_income_usd / 12.0;
    let transport_budget = input.max_payment_pct_of_income / 100.0 * monthly_gross;
    let payment_budget = (transport_budget - input.insurance_fuel_monthly_usd).max(0.0);

    let i = input.apr_pct / 100.0 / 12.0;
    let max_loan = present_value(payment_budget, i, input.loan_term_months);

    let down_frac = (input.down_payment_pct / 100.0).clamp(0.0, 0.99);
    let max_price = max_loan / (1.0 - down_frac);

    CarAffordResult {
        monthly_transport_budget_usd: transport_budget,
        monthly_payment_budget_usd: payment_budget,
        max_loan_usd: max_loan,
        max_car_price_usd: max_price,
        down_payment_needed_usd: max_price * down_frac,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 1e-2
    }

    fn base() -> CarAffordInput {
        CarAffordInput {
            annual_income_usd: 60_000.0,
            down_payment_pct: 20.0,
            loan_term_months: 48.0,
            apr_pct: 6.0,
            max_payment_pct_of_income: 10.0,
            insurance_fuel_monthly_usd: 0.0,
        }
    }

    #[test]
    fn transport_budget() {
        // 10% × 5,000 = 500.
        assert!(close(analyze(&base()).monthly_transport_budget_usd, 500.0));
    }

    #[test]
    fn max_loan() {
        assert!(close(analyze(&base()).max_loan_usd, 21290.1589));
    }

    #[test]
    fn max_price_grosses_up_for_down_payment() {
        assert!(close(analyze(&base()).max_car_price_usd, 26612.6986));
    }

    #[test]
    fn down_payment_needed() {
        let r = analyze(&base());
        assert!(close(r.down_payment_needed_usd, r.max_car_price_usd * 0.20));
    }

    #[test]
    fn insurance_fuel_reduces_budget() {
        let r = analyze(&CarAffordInput {
            insurance_fuel_monthly_usd: 200.0,
            ..base()
        });
        assert!(close(r.monthly_payment_budget_usd, 300.0));
        assert!(r.max_car_price_usd < analyze(&base()).max_car_price_usd);
    }

    #[test]
    fn higher_income_affords_more() {
        let low = analyze(&base());
        let high = analyze(&CarAffordInput {
            annual_income_usd: 120_000.0,
            ..base()
        });
        assert!(high.max_car_price_usd > low.max_car_price_usd);
    }

    #[test]
    fn zero_apr_is_payment_times_term() {
        let r = analyze(&CarAffordInput {
            apr_pct: 0.0,
            ..base()
        });
        // 500 × 48 = 24,000 loan; price = 24,000 / 0.8 = 30,000.
        assert!(close(r.max_loan_usd, 24_000.0));
        assert!(close(r.max_car_price_usd, 30_000.0));
    }

    #[test]
    fn shorter_term_lowers_loan() {
        let short = analyze(&CarAffordInput {
            loan_term_months: 36.0,
            ..base()
        });
        assert!(short.max_loan_usd < analyze(&base()).max_loan_usd);
    }
}
